gax/retry_policy.h

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_API_CODEGEN_GAPIC_RETRY_POLICY_H_
+#define GOOGLE_API_CODEGEN_GAPIC_RETRY_POLICY_H_
+
+#include <chrono>
+#include <memory>
+
+namespace google {
+namespace gax {
+
+/**
+ * Define the interface for controlling how clients retry RPC operations.
+ *
+ * Idempotent (and certain non-idempotent) operations can be retried
+ * transparently to the user. However, we need to give the users enough flexiblity
+ * to control when to stop retrying.
+ *
+ * The application provides an instance of this class when the client is created.
+ */
+template <typename StatusType, typename RetryablePolicy>
+class RetryPolicy {
+ public:
+  virtual ~RetryPolicy() = default;
+
+  /**
+   * Return a new copy of this object with the same retry criteria and fresh state.
+   */
+  virtual std::unique_ptr<RetryPolicy<StatusType, RetryablePolicy>> clone() const = 0;
+
+  /**
+   * Handle an RPC failure
+   *
+   * @return true if the RPC operation should be retried.
+   */
+  virtual bool OnFailure(StatusType const& status) = 0;
+};
+
+/**
+ * Implement a simple "count errors and then stop" retry policy.
+ */
+template <typename StatusType, typename RetryablePolicy>
+class LimitedErrorCountRetryPolicy : RetryPolicy<StatusType, RetryablePolicy> {
+ public:
+  using BaseType = RetryPolicy<StatusType, RetryablePolicy>;
```

Comment:
Is this necessary, or just to save typing? I don't think it should be public in any case.

Reply:
Just saving typing in the function body of clone. I was careful not to expose BaseType in the signature of any of the API methods but was sloppy in exposing it publicly.

Reply:
Just saves typing in clone and keeps line length more reasonable. I was careful not to expose BaseType in any method signatures but was sloppy right here.